File: typescript/tests/ulid_vectors.test.ts
/**
 * Replays the committed strict-ULID vectors against the TypeScript
 * implementation.
 *
 * The vectors live at conformance/ulid_vectors.json and are authored in
 * conformance/gen_ulid_vectors.py -- not derived from any implementation. The
 * Python and Go suites replay the same file, which is what pins the profile
 * (docs/process-trace-store.md, "Identifiers") across three independent
 * minters.
 */

import { strict as assert } from "node:assert";
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { test } from "node:test";
import { fileURLToPath } from "node:url";
import {
	ulidEncode,
	ulidMint,
	ulidTimestamp,
	ulidValid,
} from "../src/trace.js";

// Compiled test lives at typescript/dist-test/tests/, so the vectors are three
// directories up and across into conformance/.
const vectorsPath = join(
	dirname(fileURLToPath(import.meta.url)),
	"..",
	"..",
	"..",
	"conformance",
	"ulid_vectors.json",
);

interface EncodeVector {
	readonly name: string;
	readonly ms: number;
	readonly random_hex: string;
	readonly ulid: string;
}

interface ParseVector {
	readonly name: string;
	readonly text: string;
	readonly valid: boolean;
	readonly ms?: number;
}

interface VectorDoc {
	readonly encode_vector_count: number;
	readonly parse_vector_count: number;
	readonly encode_vectors: readonly EncodeVector[];
	readonly parse_vectors: readonly ParseVector[];
}

const doc = JSON.parse(readFileSync(vectorsPath, "utf8")) as VectorDoc;

test("ulid vectors: the document's own counts hold", () => {
	assert.equal(doc.encode_vectors.length, doc.encode_vector_count);
	assert.equal(doc.parse_vectors.length, doc.parse_vector_count);
	assert.ok(doc.encode_vector_count > 0);
	assert.ok(doc.parse_vector_count > 0);
});

test("ulid vectors: every encode vector reproduces byte-for-byte", () => {
	for (const vec of doc.encode_vectors) {
		const randomness = Uint8Array.from(Buffer.from(vec.random_hex, "hex"));
		assert.equal(randomness.length, 10, vec.name);
		const encoded = ulidEncode(vec.ms, randomness);
		assert.equal(encoded, vec.ulid, vec.name);
		assert.equal(encoded.length, 26, vec.name);
		// Every encoding round-trips to the millisecond it carries.
		assert.equal(ulidTimestamp(encoded), vec.ms, vec.name);
	}
});

test("ulid vectors: every parse vector yields the pinned verdict", () => {
	for (const vec of doc.parse_vectors) {
		const ms = ulidTimestamp(vec.text);
		if (vec.valid) {
			assert.equal(ms, vec.ms, vec.name);
			assert.equal(ulidValid(vec.text), true, vec.name);
		} else {
			assert.equal(ms, null, vec.name);
			assert.equal(ulidValid(vec.text), false, vec.name);
		}
	}
});

test("ulid: a minted identifier carries the clock and is canonical", () => {
	const minted = ulidMint(1786594672913);
	assert.equal(minted.length, 26);
	assert.equal(ulidTimestamp(minted), 1786594672913);
	for (const ch of minted) {
		assert.ok("0123456789ABCDEFGHJKMNPQRSTVWXYZ".includes(ch));
	}
});

test("ulid: 80 crypto-random bits differ across mints", () => {
	const seen = new Set<string>();
	for (let i = 0; i < 64; i++) {
		seen.add(ulidMint(1786594672913));
	}
	assert.equal(seen.size, 64);
});

test("ulid: a non-string is never a valid identifier", () => {
	assert.equal(ulidTimestamp(null), null);
	assert.equal(ulidTimestamp(12345), null);
	assert.equal(ulidValid(undefined), false);
});

File: go/strictcli/ulid_vectors_test.go
package strictcli

// Replays the committed strict-ULID vectors against the Go implementation.
//
// The vectors live at conformance/ulid_vectors.json and are authored in
// conformance/gen_ulid_vectors.py -- not derived from any implementation. The
// Python and TypeScript suites replay the same file, which is what pins the
// profile (docs/process-trace-store.md, "Identifiers") across three
// independent minters.

import (
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"testing"
)

type ulidEncodeVector struct {
	Name      string `json:"name"`
	MS        int64  `json:"ms"`
	RandomHex string `json:"random_hex"`
	ULID      string `json:"ulid"`
}

type ulidParseVector struct {
	Name  string `json:"name"`
	Text  string `json:"text"`
	Valid bool   `json:"valid"`
	MS    int64  `json:"ms"`
}

type ulidVectorDoc struct {
	EncodeCount   int                `json:"encode_vector_count"`
	ParseCount    int                `json:"parse_vector_count"`
	EncodeVectors []ulidEncodeVector `json:"encode_vectors"`
	ParseVectors  []ulidParseVector  `json:"parse_vectors"`
}

// ulidVectorsPath locates conformance/ulid_vectors.json relative to this source
// file via runtime.Caller, so the test works regardless of the process working
// directory.
func ulidVectorsPath(t *testing.T) string {
	t.Helper()
	_, thisFile, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("runtime.Caller failed")
	}
	return filepath.Join(filepath.Dir(thisFile), "..", "..", "conformance", "ulid_vectors.json")
}

func loadULIDVectors(t *testing.T) ulidVectorDoc {
	t.Helper()
	data, err := os.ReadFile(ulidVectorsPath(t))
	if err != nil {
		t.Fatalf("reading vectors: %v", err)
	}
	var doc ulidVectorDoc
	if err := json.Unmarshal(data, &doc); err != nil {
		t.Fatalf("parsing vectors: %v", err)
	}
	if len(doc.EncodeVectors) != doc.EncodeCount || doc.EncodeCount == 0 {
		t.Fatalf("encode vector count %d does not match %d", len(doc.EncodeVectors), doc.EncodeCount)
	}
	if len(doc.ParseVectors) != doc.ParseCount || doc.ParseCount == 0 {
		t.Fatalf("parse vector count %d does not match %d", len(doc.ParseVectors), doc.ParseCount)
	}
	return doc
}

func TestULIDEncodeVectors(t *testing.T) {
	for _, vec := range loadULIDVectors(t).EncodeVectors {
		randomness, err := hex.DecodeString(vec.RandomHex)
		if err != nil {
			t.Fatalf("%s: bad random_hex: %v", vec.Name, err)
		}
		if len(randomness) != 10 {
			t.Fatalf("%s: random_hex is %d bytes, want 10", vec.Name, len(randomness))
		}
		got := ulidEncode(vec.MS, randomness)
		if got != vec.ULID {
			t.Errorf("%s: encoded %q, want %q", vec.Name, got, vec.ULID)
		}
		if len(got) != 26 {
			t.Errorf("%s: encoding is %d characters, want 26", vec.Name, len(got))
		}
		// Every encoding round-trips to the millisecond it carries.
		ms, ok := ulidTimestamp(got)
		if !ok || ms != vec.MS {
			t.Errorf("%s: round-trip yielded (%d, %v), want (%d, true)", vec.Name, ms, ok, vec.MS)
		}
	}
}

func TestULIDParseVectors(t *testing.T) {
	for _, vec := range loadULIDVectors(t).ParseVectors {
		ms, ok := ulidTimestamp(vec.Text)
		if ok != vec.Valid {
			t.Errorf("%s: parsed %q as valid=%v, want %v", vec.Name, vec.Text, ok, vec.Valid)
			continue
		}
		if vec.Valid && ms != vec.MS {
			t.Errorf("%s: parsed %q to %d, want %d", vec.Name, vec.Text, ms, vec.MS)
		}
		if ulidValid(vec.Text) != vec.Valid {
			t.Errorf("%s: ulidValid disagrees with ulidTimestamp", vec.Name)
		}
	}
}

func TestULIDMintCarriesTheClockAndIsCanonical(t *testing.T) {
	minted, err := ulidMint(1786594672913)
	if err != nil {
		t.Fatalf("mint: %v", err)
	}
	if len(minted) != 26 {
		t.Fatalf("minted %q is %d characters, want 26", minted, len(minted))
	}
	ms, ok := ulidTimestamp(minted)
	if !ok || ms != 1786594672913 {
		t.Fatalf("minted %q parsed to (%d, %v)", minted, ms, ok)
	}
}

func TestULIDMintRandomnessDiffersAcrossCalls(t *testing.T) {
	// 80 crypto-random bits: a collision in a small sample is not credible.
	seen := map[string]bool{}
	for i := 0; i < 64; i++ {
		minted, err := ulidMint(1786594672913)
		if err != nil {
			t.Fatalf("mint: %v", err)
		}
		if seen[minted] {
			t.Fatalf("minted %q twice", minted)
		}
		seen[minted] = true
	}
}
